### PR TITLE
serialization: prevent int overflow for big Coin::nHeight

### PR DESCRIPTION
**Pull request by pierreN**
This is an attempt to fix fuzzer issues 1,2,8 reported by practicalswift here : https://github.com/bitcoin/bitcoin/issues/18046

The fuzzer harness doesn't prevent deserialization of unrealistic high values for `Coin::nHeight`.  In the [provided examples](https://github.com/bitcoin/bitcoin/issues/18046), we have : 
- `blockundo_deserialize` : the varint `0x8DD88DD700` is deserialized as `3944983552` in `Coin::nHeight` (`TxInOutFormatter::Unser`)
- `coins_deserialize` : the varint `0x8DD5D5EC40` is deserialized as `3939874496` similarly
- `txundo_deserialize`: the varint `0x8DCD828F01` is deserialized as `3921725441` in `Coin::nHeight` (`Coin::Unserialize`)

Since `Coin::nHeight` is 31 bit long, multiplying a large value by 2 triggers the fuzzer.

AFAIK those values are unrealistic (~70k years for the smallest..). I've looked a bit a reducing the range of values the fuzzer can deserialize, but this seems to be too much code change for not much.

Hence this PR chooses to static cast `nHeight` when re-serializing; it seems to be the less intrusive/safest way to prevent the fuzzer output.

Another more "upstream" approach would be to limit `Coin::nHeight` values to something more realistic, e.g. `0xFFFFFFF` (~5k years) : 

https://github.com/pierreN/bitcoin/blob/de3a30bab28e2db853a795017c5ec1704a1d0fee/src/undo.h#L39 and https://github.com/pierreN/bitcoin/blob/de3a30bab28e2db853a795017c5ec1704a1d0fee/src/coins.h#L71

Thanks !

NB: i was also not sure about the component/area to prefix the PR/commit with.. ?

### CODE REVIEW

**Comment by pierreN** [comment]
https://github.com/pierreN/bitcoin/blob/de3a30bab28e2db853a795017c5ec1704a1d0fee/src/undo.h#L36-L40

**Comment by practicalswift** [comment]
@pierreN 

Thanks for tackling this.

Consider throwing a `std::ios_base::failure` instead of silently truncating in case of serialization/deserialization errors.

That is the pattern used generally in the serialization/deserialisation code.

**Comment by practicalswift** [comment]
@pierreN 

FWIW, this is what I use in my local tree to avoid hitting the signed integer overflow when fuzzing:

```diff
diff --git a/src/coins.h b/src/coins.h
index e71c8a47b..2be607207 100644
--- a/src/coins.h
+++ b/src/coins.h
@@ -59,7 +59,7 @@ public:
     template<typename Stream>
     void Serialize(Stream &s) const {
         assert(!IsSpent());
-        uint32_t code = nHeight * 2 + fCoinBase;
+        const uint32_t code = nHeight * uint32_t{2} + fCoinBase;
         ::Serialize(s, VARINT(code));
         ::Serialize(s, Using<TxOutCompression>(out));
     }
@@ -68,7 +68,11 @@ public:
     void Unserialize(Stream &s) {
         uint32_t code = 0;
         ::Unserialize(s, VARINT(code));
-        nHeight = code >> 1;
+        const uint32_t block_height = code >> 1;
+        if (block_height > (std::numeric_limits<uint32_t>::max() - 1) / 2) {
+            throw std::ios_base::failure("Invalid nHeight when deserializing Coin.");
+        }
+        nHeight = block_height;
         fCoinBase = code & 1;
         ::Unserialize(s, Using<TxOutCompression>(out));
     }
```

Feel free to use it - I'd love to see it upstreamed. That would simplify my fuzzing workflow :)

**Comment by sipa** [comment]
Agree with throwing an error when the value is out of range.

**Comment by pierreN** [comment]
OK, thanks.

Note that strictly speaking, as long as [`unsigned int`](https://github.com/bitcoin/bitcoin/blob/e3154aacf444a29f9f5642b6c937e875ab9a2c13/src/undo.h#L37) and `uint32_t` are 32b long, in the code posted by @practicalswift only the `uint32_t{2}` are necessary for the fuzzer exemples (to the compiler this is equivalent to the `static_cast` in this PR which didn't truncate them - sorry for being unclear).

I updated the branch to use the `uint32_t{}` and made explicit the `unsigned int` == `uint32_t`.

But AFAIK exceptions are already raised in `serialize.h:ReadVarInt` ([here](https://github.com/bitcoin/bitcoin/blob/e3154aacf444a29f9f5642b6c937e875ab9a2c13/src/serialize.h#L461) and [here](https://github.com/bitcoin/bitcoin/blob/e3154aacf444a29f9f5642b6c937e875ab9a2c13/src/serialize.h#L466)) : I don't think you can reach the new `Invalid nHeight when deserializing Coin` exception.

Examples :
- `code = VARINT(0x8e fe fe fe 7f)` will evaluate `block_height > (std::numeric_limits<uint32_t>::max() - 1) / 2)` to `false`
- `code = VARINT(0x8e fe fe ff 00)` will not fit an `uint32_t` hence raise [this](https://github.com/bitcoin/bitcoin/blob/e3154aacf444a29f9f5642b6c937e875ab9a2c13/src/serialize.h#L461) already existing exception

Did I miss something by any chance ?

**Comment by MarcoFalke** [comment]
ACK 21379af7537c71181dab0e9de3af5459a0f4a00e 📀

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 21379af7537c71181dab0e9de3af5459a0f4a00e 📀
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUj3qwv7BGTYcciiJxr/pawLmVnsWE3zYTUPkpxF6yHvtWRJUaXA7ZfPs9CHsHCI
1bNzPhPSel9+IpoZ5xx71UN0yNUPBxw1ZQ4PhnEffRXsdhWEz3BjhORZ+IJ+hFVq
xx1LPZjRjd97WARFzOnsP98ayBFFxIh9bSlA0U8jKm8BAFnONrqpI1v5CMBxdnai
OYHF5hQM1DKM4YaPT6ZOEwknV7WO8Nifu+kAysvdoI+M0E45uzf+mDlx2qyP1LYf
b0ErFwTdPY7ldDLBBqOuiV/R21cZyhQBEbXUpTCwzxWoHcwEHoSLUITJ+YQ7LvTR
ET5/d8kVsBm6o1z5mtnXrHPwXEvlKbHMFtPw6ZpCTTf5FLLZTVOC32/JVIKle7gj
a6AKmoQDEz1+XyWCHqMCBad7TnCa78k7LrGL8NAV0RoezqazyncNrri/pG0OgdKW
t49UxzN7Pewq6ApqsHyEAgkeC3DVzw5HsrdHakIYbZu096kMuEpHoZZfL0KFkl4/
XBxEu4pW
=DB20
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `0fdeb1acb1aa9d5b5bfc0fded9cbedaaabafa625d387583314adb45f1c22a61e  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401080fdeb1acb1aa9d5b5bfc0fded9cbedaaabafa625d387583314adb45f1c22a61ef010453ec514da1e6404934baba8798a9df908fff0101377a355fe716587c4e3901779bb75cf08f1045e7e401bf008b0d181cbb85e512d0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010474192ff025c91277e6ed2e21c688a2508f1045e7e401bf008b401a52662e2dd610083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0109c28e0726aa1f25452997df08526464208f1045e7e401bf00837049decca5dc8b90083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010f25701712ae4542e0c14b2de5a8b182308f1045e7e401bf0080ddb3d8b5171754b0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

File: src/undo.h [hunk 1]
```diff
@@ -34,7 +34,7 @@ struct TxInUndoFormatter
 
     template<typename Stream>
     void Unser(Stream &s, Coin& txout) {
-        unsigned int nCode = 0;
+        uint32_t nCode = 0;
```
Comment (ryanofsky) [inline src/undo.h]: Again, not sure if I understand this correctly.

The undefined behavior happens because standard only says `unsigned int nCode` has to be 16+ bits.

 If `nCode` is 32 bits everything works perfectly. Any serialized VarInt value that is 2**32 or more will trigger "ReadVarInt(): size too large" exception, and any smaller values will be broken up into height and fcoinbase correctly.

If `nCode` is less than 32 bits, ReadVarInt() exceptions will be triggered unnecessarily. If `nCode` is more than 32 bits, top bits that are set will be silently discarded without triggering the ReadVarInt() exceptions that should be triggered.

Chaning `nCode` to be 32 bits instead of having an implementation defined width gives the right behavior on all platforms.

Again, I'm wondering if this is right or something else is going on.

**Review by ryanofsky — APPROVED** [review]
Code review ACK 21379af7537c71181dab0e9de3af5459a0f4a00e assuming I'm actually understanding this PR correctly. I left some questions in review comments.

File: src/coins.h [hunk 2]
```diff
@@ -59,7 +59,7 @@ class Coin
     template<typename Stream>
     void Serialize(Stream &s) const {
         assert(!IsSpent());
-        uint32_t code = nHeight * 2 + fCoinBase;
+        uint32_t code = nHeight * uint32_t{2} + fCoinBase;
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/coins.h]: I believe we're in fact already assuming that short is 16 bits, and int is 32 bits (see src/compat/assumptions.h). Maybe should just make those assumptions more explicit (asserting that the range of int matches that of int32_t, etc., ...).

**Comment by practicalswift** [comment]
ACK 21379af7537c71181dab0e9de3af5459a0f4a00e

Would be nice to have this merged to avoid having UBSan hitting the signed integer overflow over and over when fuzzing :)

File: src/coins.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/coins.h]: > 2 operand was a 2+ bit integer with actual width being platform specific but at least 2.

Not sure, but I think you got this wrong. `2` should be an `int`, which should be promoted to `uint32_t`, if I read [this](https://en.cppreference.com/w/cpp/language/operator_arithmetic#Conversions) right. ("if the unsigned operand's conversion rank is greater or equal to the conversion rank of the signed operand, the signed operand is converted to the unsigned operand's type") and [this]() ("The rank of any unsigned type is equal to the rank of the corresponding signed type")

So this change in code should not result in any change to the compiled code. I checked locally that clang++ produces the same bitcoind for me with O2.

As a side note, the boolean is converted to `int` in a first step. See [here](https://en.cppreference.com/w/cpp/language/implicit_conversion#Integral_promotion). Then  promoted to unsigned for the same reason I mentioned above.

File: src/undo.h [hunk 3]
```diff
@@ -24,7 +24,7 @@ struct TxInUndoFormatter
 {
     template<typename Stream>
     void Ser(Stream &s, const Coin& txout) {
-        ::Serialize(s, VARINT(txout.nHeight * 2 + (txout.fCoinBase ? 1u : 0u)));
+        ::Serialize(s, VARINT(txout.nHeight * uint32_t{2} + (txout.fCoinBase ? 1u : 0u)));
```
Comment (MarcoFalke) [inline src/undo.h]:
```suggestion
        ::Serialize(s, VARINT(txout.nHeight * uint32_t{2} + txout.fCoinBase));
```

It could make sense to change this for symmetry with the implementation in `Coin::Serialize`?

File: src/undo.h (hunk 1 above)
Comment (MarcoFalke) [reply src/undo.h]: > The undefined behavior happens because standard only says unsigned int nCode has to be 16+ bits.

I don't think any undefined behavior has been observed here. This has just been changed for clarity and should also not result in a change in the compiled code.

Comment (pierreN) [reply src/undo.h]: Yes this is mostly for the sake of clarity. I'm struggling to see bitcoin run on platforms where `unsigned int` is different than 32 bits but who knows.

> If nCode is less than 32 bits, ReadVarInt() exceptions will be triggered unnecessarily.

If on an imaginary platform `unsigned int` is 16 bits, then yes `nCode` have to be less than `2**16/(6*24*365)<2 years` so this code would already hit the exception in `ReadVarInt`. 

> If nCode is more than 32 bits, top bits that are set will be silently discarded without triggering the ReadVarInt() exceptions that should be triggered.

Yes. In that case adding the new exception would be necessary (limiting the threshold to `unsigned int`, not `uint32_t`).

File: src/undo.h (hunk 3 above)
Comment (pierreN) [reply src/undo.h]: This looks like a great idea. I'm double checking it.

[the remark this answers is not in the document]
Comment (pierreN) [reply src/undo.h]: Yes, TBH I was unsure why the ternary operator was used.

File: src/coins.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (pierreN) [reply src/coins.h]: > I believe we're in fact already assuming that short is 16 bits, and int is 32 bits (see src/compat/assumptions.h).

Ow, thank you, I hadn't see this file. Those are great assumptions.

> As a side note, the boolean is converted to int in a first step. See here. Then promoted to unsigned for the same reason I mentioned above.

Yes, my basic understanding was that `2` was being understood as an `int`, forcing the `nHeight` to be implicitly casted as an `int`, forcing `nHeight * 2 + fCoinBase` to be implicitely casted as an `int` then implicitly casted to an `uint32_t`.

Hence forcing `2` (or `nHeight`) to be an `uint32_t` makes all the implicit casting disappears.
If you really wanted to be explicit, you could do `static_cast<uint32_t>{nHeight} * uint32_t{2} + static_cast<uint32_t>(fCoinBase)` but this is un-necessarily verbose..

[the remark this answers is not in the document]
Comment (ryanofsky) [reply src/coins.h]: re: https://github.com/bitcoin/bitcoin/pull/18433#discussion_r399551704

> > 2 operand was a 2+ bit integer with actual width being platform specific but at least 2.
> 
> Not sure, but I think you got this wrong. `2` should be an `int`, which should be promoted to `uint32_t`

Oh you're right it must be a 16+ bit signed int: https://en.cppreference.com/w/cpp/language/integer_literal#The_type_of_the_literal

I think it doesn't change the conclusion that the only way this change could change behavior is on the imaginary platform with 31 bit ints

File: src/undo.h (hunk 1 above)
Comment (ryanofsky) [reply src/undo.h]: re: https://github.com/bitcoin/bitcoin/pull/18433#discussion_r399552921

> I don't think any undefined behavior has been observed here. This has just been changed for clarity and should also not result in a change in the compiled code.

Good to know. Any I don't think it was right for me to call it "undefined behavior", more like "platform specific behavior" only on a theoretical platform where an int is 31 bits with when height is somehow >= 2**30.

In the future, it would be good to do fuzzer fixes and code cleanups in separate commits. Or at least say what the different changes are intended to do in the commit message.  The changes are enough work to understand without having to guess which are real and which are no-ops!

**Review by ryanofsky — APPROVED** [review]
Still looks good, thanks for explanations!

**Comment by pierreN** [comment]
Thanks for the reviews, I removed the ternary operator on `fCoinBase` as requested. 
In the same spirit, I also changed the `nCode / 2` into `nCode >> 2` to match `undo.h` and `coins.h`.

**Review by ryanofsky — APPROVED** [review]
Code review ACK e980214bc4fd49530e8d50fe0a6657b8583bc6b5. Just removed ternary ? 1 : 0 and replaced / 2 with >> 1 since last review

**Comment by MarcoFalke** [comment]
re-ACK e980214bc4fd49530e8d50fe0a6657b8583bc6b5 🎑

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK e980214bc4fd49530e8d50fe0a6657b8583bc6b5 🎑
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgKLgv+MDL+1GuBwP8B6du9vzoVuuWIcYe1V6tmN4/0EnM5llvYevJznwvpSuRM
LwfFEkQ8fgAqiJDwbzM50sjaaC6eYj36uQShVCp2DcuD1mFosnaGzX9u60xQPbpu
HY7WytxaHwRyN3XSrXHuEL2PuVR79wQgAnwqaBxZ25KY4bb3ab0jLVNJb1rfNLuX
fVoBrXdLWH0vm8ompFuhGylWu5u4gVkIO+auemVAGECIMhu2238HkODZ9aE6rjWR
FNKNiSTVeAsD9iaqZo6Tw0lyf1SLSJuuHdTtQ5vIG70r/dNPK8ba20oHC1Ss22RK
Vsie4z2Zl2wCLbqIVKQJPIw0JEOkzUwr4BnnnbzizYkOAGShQq4iIVd3UA44H+Bp
lo6cwp/nn9WgU+fdcF1zdWmArfRHhh3Ocv3ENqwfL/AuT7wGqDe0YTBb/ZWPq6Vm
XFi2RK3qRrxyC4URXx13LGclrEGNVAHGnqoNlDF+Kz4jUsv6w81+gMYqlS8qCBvx
ZlP9A9rc
=/tH2
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `4660dfc715111cfef91d5f61080473e35fd321fe8ef6e8c0530e5789db28a610  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401084660dfc715111cfef91d5f61080473e35fd321fe8ef6e8c0530e5789db28a610f01071c2b44ba3b6c033a6af411238e4165608fff0104c66dc41a783f6c8f4e934954d86182a08f1045e7e9dfaf008556d4baded70263d0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010d7668590eafc4819fc18e250374fd24008f1045e7e9dfdf00814642ee10ab398870083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010ecd1b0b08d43e838d12812435a279e5e08f1045e7e9dfaf008757e38ca64986b820083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

**Comment by sipa** [comment]
utACK e980214bc4fd49530e8d50fe0a6657b8583bc6b5

**Comment by practicalswift** [comment]
ACK e980214bc4fd49530e8d50fe0a6657b8583bc6b5 -- patch looks correct

File: src/undo.h [hunk 4]
```diff
@@ -34,9 +34,9 @@ struct TxInUndoFormatter
 
     template<typename Stream>
     void Unser(Stream &s, Coin& txout) {
-        unsigned int nCode = 0;
+        uint32_t nCode = 0;
         ::Unserialize(s, VARINT(nCode));
-        txout.nHeight = nCode / 2;
+        txout.nHeight = nCode >> 1;
```
Comment (promag) [inline src/undo.h]: Why this change?

**Review by promag — COMMENTED** [review]
ACK e980214bc4fd49530e8d50fe0a6657b8583bc6b5.

File: src/undo.h (hunk 4 above)
Comment (pierreN) [reply src/undo.h]: > In the same spirit, I also changed the nCode / 2 into nCode >> 1 to match undo.h and coins.h.

To match the code in `undo.h` and `coins.h` (as it is done with `fCoinBase`). Can remove it if people prefer ?

Comment (promag) [reply src/undo.h]: 🤷‍♂ don't make a case for me.

Comment (pierreN) [reply src/undo.h]: Hmm ok so.. Waiting for a second person opinion to remove it ?

Comment (practicalswift) [reply src/undo.h]: Agree with @promag.

Comment (MarcoFalke) [reply src/undo.h]: This pull request has 5 ACKs. Let's please not invalidate all the review and fight about style changes at this point.

Comment (promag) [reply src/undo.h]: Yeah, I wasn't trying to invalidate existing reviews, I also gave my ACK. I was trying to understand the reason behind this change. I think it's good practice to not include unrelated changes.

Comment (MarcoFalke) [reply src/undo.h]: I think it is fine to do style-changes in code that reviewers have to look at anyway. It could have been better if it was in different commits or in the same commit, but with explanation in the commit message why each change is made.
Let's just keep it in mind for the future.

Comment (pierreN) [reply src/undo.h]: My bad - I just though the change was related enough to be included here. I'll do this in a separate commit in the future. Thanks.
